File: purchases-sample-java/src/main/java/com/revenuecat/purchases/purchases_sample_java/UpsellActivity.java
package com.revenuecat.purchases.purchases_sample_java;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.SkuDetails;
import com.revenuecat.purchases.EntitlementInfo;
import com.revenuecat.purchases.Offering;
import com.revenuecat.purchases.Offerings;
import com.revenuecat.purchases.Package;
import com.revenuecat.purchases.PurchaserInfo;
import com.revenuecat.purchases.Purchases;
import com.revenuecat.purchases.PurchasesError;
import com.revenuecat.purchases.interfaces.MakePurchaseListener;
import com.revenuecat.purchases.interfaces.ReceiveOfferingsListener;
import com.revenuecat.purchases.interfaces.UpdatedPurchaserInfoListener;
import com.revenuecat.sample.R;

public class UpsellActivity extends AppCompatActivity {
    Button monthlyPurchaseView;
    Button annualPurchaseView;
    Button unlimitedPurchaseView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_upsell);

        monthlyPurchaseView = findViewById(R.id.monthly_purchase);
        annualPurchaseView = findViewById(R.id.annual_purchase);
        unlimitedPurchaseView = findViewById(R.id.unlimited_purchase);

        showScreen(null);
        findViewById(R.id.skip).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Navigator.startCatsActivity(UpsellActivity.this, false);
            }
        });
        Purchases.getSharedInstance().setUpdatedPurchaserInfoListener(new UpdatedPurchaserInfoListener() {
            @Override
            public void onReceived(@NonNull PurchaserInfo purchaserInfo) {
                checkForProEntitlement(purchaserInfo);
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        Purchases.getSharedInstance().getOfferings(new ReceiveOfferingsListener() {
            @Override
            public void onReceived(@NonNull Offerings offerings) {
                showScreen(offerings);
            }

            @Override
            public void onError(@NonNull PurchasesError error) {
                Log.e("Purchases Sample", error.getMessage());
            }
        });
    }

    private void showScreen(@Nullable Offerings offerings) {
        if (offerings == null) {
            monthlyPurchaseView.setText("Loading...");
            monthlyPurchaseView.setEnabled(false);
            annualPurchaseView.setText("Loading...");
            annualPurchaseView.setEnabled(false);
        } else {
            Offering currentOffering = offerings.getCurrent();
            if (currentOffering != null) {
                setupMonthlyButton(currentOffering.getMonthly());
                setupAnnualButton(currentOffering.getAnnual());
                setupUnlimitedButton(currentOffering.getLifetime());
            } else {
                Log.e("Purchases Sample", "Error loading current offering");
            }
        }
    }

    private void setupUnlimitedButton(@Nullable final Package lifetimePackage) {
        if (lifetimePackage != null) {
            SkuDetails unlimitedProduct = lifetimePackage.getProduct();
            String loadedText = "Buy Unlimited - " + unlimitedProduct.getPriceCurrencyCode() + " " + unlimitedProduct.getPrice();
            unlimitedPurchaseView.setTag(loadedText);
            showLoading(unlimitedPurchaseView, false);
            unlimitedPurchaseView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    makePurchase(lifetimePackage, unlimitedPurchaseView);
                }
            });
        } else {
            Log.e("Purchases Sample", "Error loading lifetime package");
        }
    }

    private void setupMonthlyButton(@Nullable final Package monthlyPackage) {
        if (monthlyPackage != null) {
            SkuDetails monthlyProduct = monthlyPackage.getProduct();
            String loadedText = "Buy Monthly - " + monthlyProduct.getPriceCurrencyCode() + " " + monthlyProduct.getPrice();
            monthlyPurchaseView.setTag(loadedText);
            showLoading(monthlyPurchaseView, false);
            monthlyPurchaseView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    makePurchase(monthlyPackage, monthlyPurchaseView);
                }
            });
        } else {
            Log.e("Purchases Sample", "Error loading monthly package");
        }
    }

    private void setupAnnualButton(@Nullable final Package annualPackage) {
        if (annualPackage != null) {
            SkuDetails annualProduct = annualPackage.getProduct();
            String loadedText = "Buy Annual - " + annualProduct.getPriceCurrencyCode() + " " + annualProduct.getPrice();
            annualPurchaseView.setTag(loadedText);
            showLoading(annualPurchaseView, false);
            annualPurchaseView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    makePurchase(annualPackage, annualPurchaseView);
                }
            });
        } else {
            Log.e("Purchases Sample", "Error loading annual package");
        }
    }

    private void makePurchase(Package packageToPurchase, final Button button) {
        showLoading(button, true);
        Purchases.getSharedInstance().purchasePackage(this, packageToPurchase, new MakePurchaseListener() {
            @Override
            public void onCompleted(@NonNull Purchase purchase, @NonNull PurchaserInfo purchaserInfo) {
                showLoading(button, false);
                checkForProEntitlement(purchaserInfo);
            }

            @Override
            public void onError(@NonNull PurchasesError error, boolean userCancelled) {
                if (!userCancelled) {
                    Log.e("Purchases Sample", error.getMessage());
                }
            }
        });
    }

    private void checkForProEntitlement(PurchaserInfo purchaserInfo) {
        EntitlementInfo proEntitlement = purchaserInfo.getEntitlements().get("pro_cat");
        if (proEntitlement != null && proEntitlement.isActive()) {
            Navigator.startCatsActivity(UpsellActivity.this, false);
        }
    }

    private void showLoading(Button button, boolean loading) {
        String text;
        if (loading) {
            text = "Loading...";
        } else {
            text = (String) button.getTag();
        }
        button.setText(text);
        button.setEnabled(!loading);
    }

}

File: purchases-sample/src/main/java/com/revenuecat/sample/UpsellActivity.kt
package com.revenuecat.sample

import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import com.revenuecat.purchases.Offerings
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaserInfo
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.getOfferingsWith
import com.revenuecat.purchases.interfaces.UpdatedPurchaserInfoListener
import com.revenuecat.purchases.purchasePackageWith
import kotlinx.android.synthetic.main.activity_upsell.annual_purchase
import kotlinx.android.synthetic.main.activity_upsell.monthly_purchase
import kotlinx.android.synthetic.main.activity_upsell.skip
import kotlinx.android.synthetic.main.activity_upsell.unlimited_purchase

class UpsellActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_upsell)

        showScreen()
        skip.setOnClickListener { startCatsActivity() }
        Purchases.sharedInstance.updatedPurchaserInfoListener =
            UpdatedPurchaserInfoListener(::checkForProEntitlement)
    }

    override fun onResume() {
        super.onResume()
        Purchases.sharedInstance.getOfferingsWith(::showError, ::showScreen)
    }

    private fun showScreen(
        offerings: Offerings? = null
    ) {
        if (offerings == null) {
            monthly_purchase.text = "Loading..."
            monthly_purchase.isEnabled = false
            annual_purchase.text = "Loading..."
            annual_purchase.isEnabled = false
        } else {
            offerings.current?.let { currentOffering ->
                setupPackageButton(currentOffering.monthly, monthly_purchase)
                setupPackageButton(currentOffering.annual, annual_purchase)
                setupPackageButton(currentOffering.lifetime, unlimited_purchase)
            } ?: showError("Error loading current offering")
        }
    }

     private fun setupPackageButton(aPackage: Package?, button: Button) {
        aPackage?.product?.let { product ->
            with(button) {
                loadedText =
                    "Buy ${aPackage.packageType} - ${product.priceCurrencyCode} ${product.price}"
                showLoading(false)
                setOnClickListener {
                    makePurchase(aPackage, this)
                }
            }
        } ?: showError("Error finding package")
    }

    private fun makePurchase(
        packageToPurchase: Package,
        button: Button
    ) {
        button.showLoading(true)
        Purchases.sharedInstance.purchasePackageWith(
            this,
            packageToPurchase,
            { error, userCancelled ->
                if (!userCancelled) {
                    showError(error)
                }
            }) { _, purchaserInfo ->
            button.showLoading(false)
            checkForProEntitlement(purchaserInfo)
        }
    }

    private fun checkForProEntitlement(purchaserInfo: PurchaserInfo) {
        if (purchaserInfo.entitlements[PREMIUM_ENTITLEMENT_ID]?.isActive == true) {
            startCatsActivity()
        }
    }

    private fun Button.showLoading(loading: Boolean) {
        text = if (loading) "Loading..." else loadedText
        isEnabled = !loading
    }
}

private var Button.loadedText: String
    get() {
        return (tag as String)
    }
    set(value) {
        tag = value
    }
